pkg/metrics/collector/netlink_metrics.go

```diff
@@ -15,85 +15,287 @@ package collector
 
 import (
 	"fmt"
+	"io/ioutil"
+	"os"
+	"strconv"
+	"sync"
 	"syscall"
 
-	"github.com/GoogleCloudPlatform/netd/pkg/tcp_metrics/inetdiag"
-	"github.com/GoogleCloudPlatform/netd/pkg/tcp_metrics/parser"
-
+	"github.com/golang/glog"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/vishvananda/netlink/nl"
+	"github.com/vishvananda/netns"
+	"golang.org/x/sys/unix"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/GoogleCloudPlatform/netd/pkg/tcp_metrics/inetdiag"
+	"github.com/GoogleCloudPlatform/netd/pkg/tcp_metrics/parser"
+	"github.com/GoogleCloudPlatform/netd/pkg/tcp_metrics/tcp"
 )
 
 var (
 	TCPConnectionsDesc = prometheus.NewDesc(
-		"tcp_connections",
+		"active_tcp_connections",
 		"Current amount of tcp connections on a node",
-		nil, nil,
+		[]string{"pod_name", "namespace_name"}, nil,
 	)
 
 	ActiveTCPRetransmits = prometheus.NewDesc(
 		"active_tcp_retransmits",
 		"Current amount of tcp retransmits on a node",
-		nil, nil,
+		[]string{"pod_name", "namespace_name"}, nil,
 	)
 )
 
+type netlinkCollector struct {
+}
+
+type netlinkStats struct {
+	retransmits    uint64
+	tcpConnections uint64
+}
+
+type podDef struct {
+	name      string
+	namespace string
+}
+
+var (
+	mapMux   sync.RWMutex
+	ipMap    = make(map[string]podDef)
+	nodeName = os.Getenv("CURRENT_NODE_NAME")
+)
+
+func init() {
+
+	glog.Infof("Initializing netlink metrics")
+	err := createPodWatch()
+	if err != nil {
+		glog.Infof("Error, pod watch not running with err, %q", err)
+	}
+
+	registerCollector("netlink_metrics", NewNetlinkCollector)
+}
+
+// Functionality for multithreaded map usage
+func safeIPWrite(key string, value podDef) {
+	mapMux.Lock()
+	defer mapMux.Unlock()
+	ipMap[key] = value
+}
+
+func safeIPDelete(key string) {
+	mapMux.Lock()
+	defer mapMux.Unlock()
+	delete(ipMap, key)
+}
+
+func safeIPRead(key string) (podDef, bool) {
+	mapMux.RLock()
+	defer mapMux.RUnlock()
+	podDef, ok := ipMap[key]
+	return podDef, ok
+}
+
+// Functionality for pod watching
+func createPodWatch() error {
+	// creates the in-cluster config
+	config, err := rest.InClusterConfig()
```

Comment:
I originally wanted to say that this may make kube-proxy/dataplane v2 a new dependency of netd because this `InClusterConfig` uses `default/kubernetes` service's cluster IP by default. Then I realized we intentionally override `KUBERNETES_SERVICE_HOST` to the actual master IP during the deployment. So this should just work :)

Reply:
Cool!

Reply:
+1 - we added that service host IP because the init container also makes a call to the master (to get the current node's pod cidr and write the ipam section of the CNI config file)